### PR TITLE
Jit pre save hook

### PR DESCRIPTION
**Pull request by mingxu1067**
### PR types

New features
### PR changes

APIs
### Describe

Pre-saving Hooks in `paddle.jit.save`

- Added pre-saving hooks mechanism to allow users or developers hand functions to preprocess `nn.Layer` before `jit.save`.
- All submitted functions are *threading-safe* and *unique-hook guaranteed*.
- Added three new APIs
1. register_save_pre_hook: 
> Register a function to _pre_save_hooks
> Args:
> - func (py-function): This function would be hanged to hooks and executed before paddle.jit.save

> Return:
> - HookRemoveHandler: A handler to remove the registered hook.
2. clear_save_pre_hooks
> Clear all hooks.

- The functions be hanged should following the interface
> def func(layer, input_spec, configs)
> - layer (`nn.Layer`): The layer passed to `jit.save`.
> - input_spec (list of `InputSpec`): The input_spec passed to `jit.save`.
> - configs (dict): The kwargs passed to `jit.save`, which would be collected into `configs`.

Usage Example:
```python
  _counter = 0
  def fake_func(layer, input_spec, configs):
      global _counter
      _counter += 1

  remove_handler = register_save_pre_hook(fake_func)
  # len(paddle.fluid.dygraph.jit._save_pre_hooks) == 1
  
  remove_handler = register_save_pre_hook(fake_func)
  # len(paddle.fluid.dygraph.jit._save_pre_hooks) == 1
  # Avoid redundancy hanging
  
  paddle.jit.save(my_layer)
  # _counter == 1
  
  paddle.jit.save(my_layer)
  # _counter == 2

  remove_handler.remove()
  # len(paddle.fluid.dygraph.jit._save_pre_hooks) == 0

  paddle.jit.save(my_layer)
  # _counter == 2

  remove_handler = register_save_pre_hook(fake_func)
  clear_save_pre_hooks()
  # len(paddle.fluid.dygraph.jit._save_pre_hooks) == 0
```

### CODE REVIEW

File: python/paddle/fluid/dygraph/jit.py [hunk 1]
```diff
@@ -525,6 +528,58 @@ def _build_load_path_and_config(path, config):
...
+    """ A HookRemoveHelper that can be used to remove hook. """
+
+    def __init__(self, hook):
+        self._hook = hook
+
+    def remove(self):
+        remove_save_pre_hook(self._hook)
+
+
+def register_save_pre_hook(hook):
```
Comment (chenwhql) [inline python/paddle/fluid/dygraph/jit.py]: we should append api doc like other api

File: python/paddle/fluid/dygraph/jit.py [hunk 2]
```diff
@@ -525,6 +528,58 @@ def _build_load_path_and_config(path, config):
...
+    global _save_pre_hooks_lock
+    global _save_pre_hooks
+    _save_pre_hooks_lock.acquire()
+    if hook not in _save_pre_hooks:
+        _save_pre_hooks.append(hook)
+    _save_pre_hooks_lock.release()
+    return HookRemoveHelper(hook)
+
+
+def remove_save_pre_hook(hook):
```
Comment (chenwhql) [inline python/paddle/fluid/dygraph/jit.py]: if `remove_save_pre_hook` is not used as public api, we recommoned nameing it with prefix `_` ,  `_remove_save_pre_hook`

File: python/paddle/fluid/dygraph/jit.py [hunk 3]
```diff
@@ -525,6 +528,58 @@ def _build_load_path_and_config(path, config):
...
+    global _save_pre_hooks_lock
+    global _save_pre_hooks
+    _save_pre_hooks_lock.acquire()
+    if hook in _save_pre_hooks:
+        _save_pre_hooks.remove(hook)
+    _save_pre_hooks_lock.release()
+
+
+def clear_save_pre_hooks():
```
Comment (chenwhql) [inline python/paddle/fluid/dygraph/jit.py]: `clear_save_pre_hooks` used in what cases? Is `HookRemoveHelper.remove` enough?

File: python/paddle/fluid/dygraph/jit.py (hunk 1 above)
Comment (mingxu1067) [reply python/paddle/fluid/dygraph/jit.py]: Done, added api doc to public functions.

File: python/paddle/fluid/dygraph/jit.py (hunk 2 above)
Comment (mingxu1067) [reply python/paddle/fluid/dygraph/jit.py]: Done, renamed `remove_save_pre_hook` to `_remove_save_pre_hook `

File: python/paddle/fluid/dygraph/jit.py (hunk 3 above)
Comment (mingxu1067) [reply python/paddle/fluid/dygraph/jit.py]: Currently not, but this is a convenient API to let user clear all hooks in one call.

Comment (chenwhql) [reply python/paddle/fluid/dygraph/jit.py]: If the specific usage scenario is not clear, we recommended to use it as an internal method first, and then upgrade it to a public API if necessary in the future

File: python/paddle/fluid/dygraph/jit.py [hunk 4]
```diff
@@ -525,6 +528,143 @@ def _build_load_path_and_config(path, config):
...
+def _remove_save_pre_hook(hook):
+    global _save_pre_hooks_lock
+    global _save_pre_hooks
+    _save_pre_hooks_lock.acquire()
+    if hook in _save_pre_hooks:
+        _save_pre_hooks.remove(hook)
+    _save_pre_hooks_lock.release()
+
+
+def run_save_pre_hooks(func):
```
Comment (chenwhql) [inline python/paddle/fluid/dygraph/jit.py]: same as `_remove_save_pre_hook`

Comment (mingxu1067) [reply python/paddle/fluid/dygraph/jit.py]: Done

File: python/paddle/fluid/dygraph/jit.py (hunk 3 above)
Comment (mingxu1067) [reply python/paddle/fluid/dygraph/jit.py]: Done, currently rename `clear_pre_save_hooks`  with prefix `_` and make it internal use only.

**Review by chenwhql — APPROVED** [review]
LGTM

**Review by TCChenlong — APPROVED** [review]

**Review by YuanRisheng — APPROVED** [review]

**Pull request merged**